extensions/cl_ext_relaxed_printf_string_address_space.asciidoc

```diff
@@ -0,0 +1,123 @@
+cl_ext_relaxed_printf_address_space
+====================================
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:blank: pass:[ +]
+:data-uri:
+:icons: font
+include::../config/attribs.txt[]
+:source-highlighter: coderay
+
+== Name Strings
+
++cl_ext_relaxed_printf_address_space+
+
+== Contact
+
+Pekka Jääskeläinen, Parmance (pekka /at/ parmance /dot/ com)
+
+== Contributors
+
+Pekka Jääskeläinen, Parmance +
+Henry Linjamäki, Parmance +
+Brice Videau, Argonne National Laboratory +
+Anastasia Stulova, Arm
+
+== Notice
+
+Copyright (c) 2022 Parmance and Argonne National Laboratory.
+
+== Status
+
+Early Draft
+
+== Version
+
+Built On: {docdate} +
+Revision: 3
+
+== Dependencies
+
+This extension is written against the OpenCL Specification Version 1.2, Revision 19 and
+the OpenCL C 2.0 Specification, Revision 29.
+
+The extension works with OpenCL 1.2 for the overloads that do not require generic address
+space support, and requires at least OpenCL 2.0 for the generic address space one.
+
+== Overview
+
+The printf as originally defined in OpenCL 1.2 deviates from the C99 printf
+enough that makes using it as a drop-in function for programs originally
+written in C99 that use printf sometimes cumbersome.
+
+This extension addresses a major gap in this regard by allowing the format
+string specifier and the arguments to the %s conversion specifier to originate
+from any address space, not only from the constant address space as specified
+in the OpenCL 1.2 specification.
+
+This relaxation implies that the format strings and the strings passed as %s
+arguments can be dynamically produced during the execution time of the kernel.
+
+== Modifications to the OpenCL C 1.2 Specification
+
+=== Modify Table 6.21
+
+Add new function overloads (added to the bottom left cell of Table 6.21): ::
++
+
+[source,c]
+----
+int printf(global char * restrict format, ...)
+int printf(local char * restrict format, ...)
+int printf(private char * restrict format, ...)
```

Comment:
These should have the `const` qualifier, i.e.:

```
int printf(global const char * restrict format, ...)
int printf(local const char * restrict format, ...)
int printf(private const char * restrict format, ...)
```

As a side-note, the existing _constant address-space_ overload is as follows:

```
int printf(constant char * restrict format, ...)
```

It would be good to add an overload for `constant const`:

```
int printf(constant const char * restrict format, ...)
```

Otherwise, adding the `const` qualifier to a string in the `constant` address space may produce a warning, e.g.:
> warning: passing argument 1 of 'printf' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
> note: expected 'constant char *' but argument is of type 'constant const char *'.

(In my own opinion, as this _arguably_ doesn't affect semantics, this we could make this change as a fix to the OpenCL spec change, rather than adding an extra overload in this extension.  If others do not share that opinion, e.g. if there is a stronger affect on semantics than I have understood, then it would seem better to add an extra overload here than to leave this unchanged, though.)